sound/soc/intel/boards/sof_sdw.c

```diff
@@ -1222,6 +1222,15 @@ static int mc_probe(struct platform_device *pdev)
 	if (!card->components)
 		return -ENOMEM;
 
+	if (mach->mach_params.dmic_num) {
```

Comment:
how about moving this `if` 7 lines up and having two different calls to `devm_kasprintf()`? I know we're about to count our PC RAM in petabytes but the old grumpy me still notes when 20 bytes are needlessly wasted even if "only" for the lifetime of the driver

Reply:
this pattern is already used in existing code, see e.g. sof_sdw_rtXYZ.c. I agree that in this case we could have done better but it's not really optimizing much so I preferred to remain consistent with the rest of the code.

Reply:
Sorry, cannot say I'm convinced. This seems like a managed resource allocation API abuse to me.

Reply:
It's not great indeed but like I said it's used in several places:

````
sof_sdw.c:	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
sof_sdw.c-					  "cfg-spk:%d cfg-amp:%d",
--
sof_sdw_max98373.c:	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
sof_sdw_max98373.c-					  "%s spk:mx8373",
--
sof_sdw_rt1308.c:	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
sof_sdw_rt1308.c-					  "%s spk:rt1308",
--
sof_sdw_rt1316.c:	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
sof_sdw_rt1316.c-					  "%s spk:rt1316",
--
sof_sdw_rt5682.c:	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
sof_sdw_rt5682.c-					  "%s hs:rt5682",
--
sof_sdw_rt700.c:	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
sof_sdw_rt700.c-					  "%s hs:rt700",
--
sof_sdw_rt711.c:	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
sof_sdw_rt711.c-					  "%s hs:rt711",
--
sof_sdw_rt711_sdca.c:	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
sof_sdw_rt711_sdca.c-					  "%s hs:rt711-sdca",
--
sof_sdw_rt715.c:	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
sof_sdw_rt715.c-					  "%s mic:rt715",
--
sof_sdw_rt715_sdca.c:	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
sof_sdw_rt715_sdca.c-					  "%s mic:rt715-sdca",
````

Optimizing the use of cfg-mics and mic information still leaves the other cases open.

Maybe use an explicit devm_kfree on the existing card->components? That would only keep what's needed allocated, but is it worth it? SoundWire only works on CML+...

Reply:
but all those examples only have one `devm_kasprintf()`, so, they aren't re-allocating memory? I meant the following:
```
	if (!mach->mach_params.dmic_num)
		card->components = devm_kasprintf(card->dev, GFP_KERNEL,
						  "cfg-spk:%d cfg-amp:%d",
						  (sof_sdw_quirk & SOF_SDW_FOUR_SPK)
						  ? 4 : 2, amp_num);
	else
		card->components = devm_kasprintf(card->dev, GFP_KERNEL,
						  "cfg-spk:%d cfg-amp:%d mic:dmic cfg-mics:%d",
						  (sof_sdw_quirk & SOF_SDW_FOUR_SPK)
						  ? 4 : 2, amp_num,
						  mach->mach_params.dmic_num);
	if (!card->components)
		return -ENOMEM;
```

Reply:
They are all re-allocating memory, see below with more lines included


````
sof_sdw.c:	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
sof_sdw.c-					  "cfg-spk:%d cfg-amp:%d",
sof_sdw.c-					  (sof_sdw_quirk & SOF_SDW_FOUR_SPK)
sof_sdw.c-					  ? 4 : 2, amp_num);
--
sof_sdw.c:		card->components = devm_kasprintf(card->dev, GFP_KERNEL,
sof_sdw.c-						  "%s mic:dmic cfg-mics:%d",
sof_sdw.c-						  card->components,
sof_sdw.c-						  mach->mach_params.dmic_num);
--
sof_sdw_max98373.c:	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
sof_sdw_max98373.c-					  "%s spk:mx8373",
sof_sdw_max98373.c-					  card->components);
--
sof_sdw_rt1308.c:	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
sof_sdw_rt1308.c-					  "%s spk:rt1308",
sof_sdw_rt1308.c-					  card->components);
--
sof_sdw_rt1316.c:	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
sof_sdw_rt1316.c-					  "%s spk:rt1316",
sof_sdw_rt1316.c-					  card->components);
--
sof_sdw_rt5682.c:	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
sof_sdw_rt5682.c-					  "%s hs:rt5682",
sof_sdw_rt5682.c-					  card->components);
--
sof_sdw_rt700.c:	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
sof_sdw_rt700.c-					  "%s hs:rt700",
sof_sdw_rt700.c-					  card->components);
--
sof_sdw_rt711.c:	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
sof_sdw_rt711.c-					  "%s hs:rt711",
sof_sdw_rt711.c-					  card->components);
--
sof_sdw_rt711_sdca.c:	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
sof_sdw_rt711_sdca.c-					  "%s hs:rt711-sdca",
sof_sdw_rt711_sdca.c-					  card->components);
--
sof_sdw_rt715.c:	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
sof_sdw_rt715.c-					  "%s mic:rt715",
sof_sdw_rt715.c-					  card->components);
--
sof_sdw_rt715_sdca.c:	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
sof_sdw_rt715_sdca.c-					  "%s mic:rt715-sdca",
sof_sdw_rt715_sdca.c-					  card->components);
````

Reply:
So that means that all those sof_sdw_*.c board drivers re-allocate the `.components` string originally allocated by sof_sdw.c? Ok, got it now, thanks!..